lib/elixir/lib/calendar/duration.ex

```diff
@@ -197,4 +197,144 @@ defmodule Duration do
       microsecond: {-ms, p}
     }
   end
+
+  @doc """
+  Parses an ISO 8601-2 formatted duration string to a `Duration` struct.
+
+  ## Examples
+
+      iex> Duration.from_iso8601("P1Y2M3DT4H5M6S")
+      {:ok, %Duration{year: 1, month: 2, day: 3, hour: 4, minute: 5, second: 6}}
+      iex> Duration.from_iso8601("PT10H30M")
+      {:ok, %Duration{hour: 10, minute: 30, second: 0}}
+      iex> Duration.from_iso8601("P3Y-2MT3H")
+      {:ok, %Duration{year: 3, month: -2, hour: 3}}
+      iex> Duration.from_iso8601("-P3Y2MT3H")
+      {:ok, %Duration{year: -3, month: -2, hour: -3}}
+      iex> Duration.from_iso8601("-P3Y-2MT3H")
+      {:ok, %Duration{year: -3, month: 2, hour: -3}}
+
+  """
+  @spec from_iso8601(String.t()) :: {:ok, t} | {:error, String.t()}
+  def from_iso8601("P" <> duration_string) do
+    parse(duration_string, %{}, "", false)
+  end
+
+  def from_iso8601("-P" <> duration_string) do
+    case parse(duration_string, %{}, "", false) do
+      {:ok, duration} ->
+        {:ok, negate(duration)}
+
+      error ->
+        error
+    end
+  end
+
+  def from_iso8601(_) do
+    {:error, "invalid duration string"}
+  end
+
+  @doc """
+  Same as parse/1 but raises an ArgumentError.
+
+  ## Examples
+
+      iex> Duration.from_iso8601!("P1Y2M3DT4H5M6S")
+      %Duration{year: 1, month: 2, day: 3, hour: 4, minute: 5, second: 6}
+
+  """
+  @spec from_iso8601!(String.t()) :: t
+  def from_iso8601!(duration_string) do
+    case from_iso8601(duration_string) do
+      {:ok, duration} ->
+        duration
+
+      {:error, reason} ->
+        raise ArgumentError, "failed to parse duration. reason: #{inspect(reason)}"
+    end
+  end
+
+  defp parse(<<>>, duration, "", _), do: {:ok, new!(Enum.into(duration, []))}
+
+  defp parse(<<c::utf8, rest::binary>>, duration, buffer, is_time)
+       when c in ?0..?9 or c in [?., ?-] do
+    parse(rest, duration, <<buffer::binary, c::utf8>>, is_time)
+  end
+
+  defp parse(<<"Y", rest::binary>>, duration, buffer, false) do
+    parse(:year, rest, duration, buffer, false)
+  end
+
+  defp parse(<<"M", rest::binary>>, duration, buffer, false) do
+    parse(:month, rest, duration, buffer, false)
+  end
+
+  defp parse(<<"W", rest::binary>>, duration, buffer, false) do
+    parse(:week, rest, duration, buffer, false)
+  end
+
+  defp parse(<<"D", rest::binary>>, duration, buffer, false) do
+    parse(:day, rest, duration, buffer, false)
+  end
+
+  defp parse(<<"T", _::binary>>, _duration, _, true) do
+    {:error, "time delimiter was already provided"}
+  end
+
+  defp parse(<<"T", rest::binary>>, duration, _buffer, false) do
+    parse(rest, duration, "", true)
+  end
```

Comment:
Instead of passing a boolean as last argument, it is easier to split these in two functions, date_parse and time_parse, and when you are in `time_parse`, you know by definition you have already seen the "T".

Reply:
What i came up with feels very verbose 😅 https://github.com/elixir-lang/elixir/pull/13473/commits/6ff56b89b955a3b332914fcfb4f5ea7f87e6d78d there's probably a neater way to go about it, let me know if you had something else in mind